### PR TITLE
feat: "deno lint" subcommand

### LINKED ISSUE #1880: Add "deno lint" subcommand
For linting files.

Probably should be implemented similar to how we've done `--fmt` ... that is get eslint into deno_std, and then load from URL in core.

### PR DESCRIPTION
**Pull request by bartlomieju**
Ref https://github.com/denoland/deno_lint/issues/78
Closes #1880

### CODE REVIEW

File: cli/flags.rs [hunk 1]
```diff
@@ -878,6 +894,18 @@ Show documentation for runtime built-ins:
...
+fn lint_subcommand<'a, 'b>() -> App<'a, 'b> {
+  SubCommand::with_name("lint")
+    .arg(unstable_arg())
+    .arg(
+      Arg::with_name("files")
+        .takes_value(true)
+        .required(true)
+        .min_values(1),
+    )
+    .about("Lint source files")
```
Comment (bartlomieju) [inline cli/flags.rs:906]: add better help message; let know about `deno-lint-ignore` comments

File: cli/main.rs [hunk 2]
```diff
@@ -312,6 +312,51 @@ async fn install_command(
...
+
+    let mut linter = deno_lint::linter::Linter::default();
+    let lint_rules = deno_lint::rules::get_all_rules();
+
+    let file_diagnostics = linter.lint(file, source_code, lint_rules)?;
+
+    error_counts += file_diagnostics.len();
+    for d in file_diagnostics.iter() {
+      eprintln!("{}", d.to_pretty_string());
+    }
```
Comment (nayeemrmn) [inline cli/main.rs:349]: Regarding diagnostic formatting, ref https://github.com/denoland/deno_lint/pull/62.

In #4832 I made syntax errors and TS type errors look like JS runtime `error.stack` (plus the source line under the message).

<kbd></kbd>

<kbd></kbd>

The result has been nice. The usage range of this format covers every kind of user error and extends naturally to `deno lint` IMO.

I think it's fine if the `deno_lint` self-contained binary looks like it currently does, but this call to `d.to_pretty_string()` should be replaced by a call to the formatting logic in `cli/diagnostics.rs`.

Comment (bartlomieju) [reply cli/main.rs:349]: @nayeemrmn thanks for pointing this out. Definitely makes sense and `deno_lint` diagnostics were actually a bit problematic as they introduce potential for dependency duplication. I'll rework the PR to use Deno's diagnostic. BTW I don't think there we'll be self-contained linter binary. It looks more like a library to build your own linter ;)

File: Cargo.lock [hunk 3]
```diff
@@ -502,6 +503,23 @@ dependencies = [
  "url 2.1.1",
 ]
 
+[[package]]
+name = "deno_lint"
+version = "0.1.6"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "98a1423991b250b313ea7d2cbe6b04dbf1235cd531ccaec32912e68fa41e1f0b"
+dependencies = [
+ "clap",
```
Comment (ry) [inline Cargo.lock:512]: Can you make it so deno_lint does not depend on clap?

Comment (bartlomieju) [reply Cargo.lock:512]: Done

**Comment by bartlomieju** [comment]
Current output using error formatting facilities from CLI as suggested by @nayeemrmn Screenshot 2020-06-07 at 22 59 31

File: cli/fmt_errors.rs [hunk 4]
```diff
@@ -10,6 +10,15 @@ use std::ops::Deref;
...
 const SOURCE_ABBREV_THRESHOLD: usize = 150;
 
+pub fn format_location(filename: String, line: i64, col: i64) -> String {
+  format!(
+    "{}:{}:{}",
+    colors::cyan(filename),
+    colors::yellow(line.to_string()),
+    colors::yellow(col.to_string())
+  )
+}
```
Comment (ry) [inline cli/fmt_errors.rs:20]: It’s unfortunate that we have two implementations of format_location.. but I guess the other is in JS

File: cli/flags.rs (hunk 1 above)
Comment (ry) [reply cli/flags.rs:906]: 👍

**Comment by bartlomieju** [comment]
We should probably support running `deno lint` without arguments - just like `deno fmt` searches for default files. I guess we'll iterate on linter a bit more before releasing 1.1 (PTAL at 0.2.0 milestone in deno_lint https://github.com/denoland/deno_lint/issues?q=is%3Aopen+is%3Aissue+milestone%3A0.2.0)

File: cli/flags.rs [hunk 5]
```diff
@@ -259,6 +262,8 @@ pub fn flags_from_vec_safe(args: Vec<String>) -> clap::Result<Flags> {
     upgrade_parse(&mut flags, m);
   } else if let Some(m) = matches.subcommand_matches("doc") {
     doc_parse(&mut flags, m);
+  } else if let Some(m) = matches.subcommand_matches("lint") {
+    lint_parse(&mut flags, m);
```
Comment (yasinaydinnet) [inline cli/flags.rs:266]: I don't know Rust, but is there an early return in Rust? If there is, wouldn't it be better if the file was something like:
```rust
...
  if let Some(m) = matches.subcommand_matches("doc") {
    return doc_parse(&mut flags, m);
  }
  if let Some(m) = matches.subcommand_matches("lint") {
    return lint_parse(&mut flags, m);
  }
```

Comment (bartlomieju) [reply cli/flags.rs:266]: This method returns `Result<Flags>` while each of the methods returns `()`.

Comment (yasinaydinnet) [reply cli/flags.rs:266]: Got it. My focus was not on the returning part but on possibility of eliminating this many `if-else`s :) Perhaps a switch statement? 
Again, this is totally trivial and unrelated. I was just looking at that file generally. Awesome work btw.

**Pull request merged**